File: src/docs/api-client-examples/javascript/tourist-workflow.js
/**
 * Tourist Hub API - Tourist Workflow Examples
 * 
 * This example demonstrates common workflows for Tourist users,
 * including profile management, tour registration, and document handling.
 */

const TouristHubClient = require('./basic-client');
const fs = require('fs');
const path = require('path');

class TouristWorkflow {
  constructor(baseURL) {
    this.client = new TouristHubClient(baseURL);
  }
  
  /**
   * Complete tourist registration and setup workflow
   */
  async registerAndSetupProfile(registrationData) {
    try {
      console.log('🚀 Starting tourist registration workflow...');
      
      // 1. Register new tourist account
      console.log('📝 Registering new account...');
      const registrationResponse = await this.client.post('/api/auth/register', {
        ...registrationData,
        userType: 'Tourist'
      });
      console.log('✅ Registration successful');
      
      // 2. Login with new credentials
      console.log('🔐 Logging in...');
      await this.client.login(registrationData.emailAddress, registrationData.password);
      
      // 3. Update profile with additional information
      console.log('👤 Updating profile...');
      const profileUpdate = {
        passportNumber: registrationData.passportNumber,
        dateOfBirth: registrationData.dateOfBirth,
        gender: registrationData.gender
      };
      
      const updatedUser = await this.client.put(`/api/users/${this.client.user.userId}`, profileUpdate);
      console.log('✅ Profile updated successfully');
      
      return updatedUser;
    } catch (error) {
      console.error('❌ Registration workflow failed:', error.message);
      throw error;
    }
  }
  
  /**
   * Browse and search available tour templates
   */
  async browseTourTemplates(filters = {}) {
    try {
      console.log('🔍 Browsing tour templates...');
      
      const templates = await this.client.get('/api/tour-templates', filters);
      
      console.log(`📋 Found ${templates.length} tour templates:`);
      templates.forEach(template => {
        console.log(`  • ${template.templateName} (${template.type})`);
        console.log(`    📅 ${template.startDate} to ${template.endDate}`);
        console.log(`    📍 ${template.sitesToVisit?.length || 0} sites to visit`);
        console.log('');
      });
      
      return templates;
    } catch (error) {
      console.error('❌ Failed to browse templates:', error.message);
      throw error;
    }
  }
  
  /**
   * View available tour events based on templates
   */
  async viewAvailableTourEvents(filters = {}) {
    try {
      console.log('🎯 Viewing available tour events...');
      
      const events = await this.client.get('/api/tour-events', filters);
      
      console.log(`🎪 Found ${events.length} available tour events:`);
      events.forEach(event => {
        console.log(`  • ${event.customTourName}`);
        console.log(`    📅 ${event.startDate} to ${event.endDate}`);
        console.log(`    👥 ${event.remainingTourists}/${event.numberOfAllowedTourists} spots available`);
        console.log(`    🏨 Hotels: ${event.place1Hotel}, ${event.place2Hotel}`);
        console.log(`    📊 Status: ${event.status}`);
        console.log('');
      });
      
      return events;
    } catch (error) {
      console.error('❌ Failed to view tour events:', error.message);
      throw error;
    }
  }
  
  /**
   * Register for a tour event
   */
  async registerForTour(tourEventId) {
    try {
      console.log(`📝 Registering for tour event: ${tourEventId}`);
      
      // First, get tour event details
      const tourEvent = await this.client.get(`/api/tour-events/${tourEventId}`);
      console.log(`🎪 Tour: ${tourEvent.customTourName}`);
      
      // Check availability
      if (tourEvent.remainingTourists <= 0) {
        throw new Error('Tour is fully booked');
      }
      
      if (tourEvent.status !== 'Active') {
        throw new Error(`Tour is not available for registration (Status: ${tourEvent.status})`);
      }
      
      // Register for the tour
      const registration = await this.client.post(`/api/tour-events/${tourEventId}/register`);
      
      console.log('✅ Registration successful!');
      console.log(`📋 Registration Status: ${registration.status}`);
      
      if (registration.status === 'Pending') {
        console.log('⏳ Your registration is pending approval from the provider');
      }
      
      return registration;
    } catch (error) {
      console.error('❌ Tour registration failed:', error.message);
      throw error;
    }
  }
  
  /**
   * View tour schedule and activities
   */
  async viewTourSchedule(tourEventId) {
    try {
      console.log(`📅 Viewing schedule for tour: ${tourEventId}`);
      
      const schedule = await this.client.get(`/api/tour-events/${tourEventId}/schedule`);
      
      console.log('📋 Tour Schedule:');
      schedule.forEach(activity => {
        console.log(`  📅 ${activity.activityDate} (${activity.islamicDate || 'N/A'})`);
        console.log(`    🕐 ${activity.startTime} - ${activity.endTime}`);
        console.log(`    🎯 ${activity.activityType}: ${activity.description}`);
        console.log(`    📍 Location: ${activity.location}`);
        if (activity.webLink) {
          console.log(`    🔗 Link: ${activity.webLink}`);
        }
        console.log('');
      });
      
      return schedule;
    } catch (error) {
      console.error('❌ Failed to view schedule:', error.message);
      throw error;
    }
  }
  
  /**
   * Upload document (passport, ticket, etc.)
   */
  async uploadDocument(filePath, documentType, description = '') {
    try {
      console.log(`📄 Uploading ${documentType} document...`);
      
      // Check if file exists
      if (!fs.existsSync(filePath)) {
        throw new Error(`File not found: ${filePath}`);
      }
      
      // Create file object (in browser, this would be a File object)
      const fileBuffer = fs.readFileSync(filePath);
      const fileName = path.basename(filePath);
      
      // In a real browser environment, you would use:
      // const file = document.getElementById('fileInput').files[0];
      
      const uploadData = {
        type: documentType,
        description: description
      };
      
      // Note: This is a simplified example. In practice, you'd need to handle
      // file uploads differently in Node.js vs browser environments
      const document = await this.client.uploadFile('/api/documents', {
        buffer: fileBuffer,
        originalname: fileName,
        mimetype: this.getMimeType(fileName)
      }, uploadData);
      
      console.log('✅ Document uploaded successfully');
      console.log(`📄 Document ID: ${document.documentId}`);
      console.log(`📁 File: ${document.fileName}`);
      
      return document;
    } catch (error) {
      console.error('❌ Document upload failed:', error.message);
      throw error;
    }
  }
  
  /**
   * View uploaded documents
   */
  async viewMyDocuments() {
    try {
      console.log('📂 Viewing my documents...');
      
      const documents = await this.client.get('/api/documents');
      
      console.log(`📋 Found ${documents.length} documents:`);
      documents.forEach(doc => {
        console.log(`  📄 ${doc.fileName} (${doc.type})`);
        console.log(`    📝 ${doc.description || 'No description'}`);
        console.log(`    📅 Uploaded: ${new Date(doc.uploadDate).toLocaleDateString()}`);
        console.log(`    💾 Size: ${this.formatFileSize(doc.fileSize)}`);
        console.log('');
      });
      
      return documents;
    } catch (error) {
      console.error('❌ Failed to view documents:', error.message);
      throw error;
    }
  }
  
  /**
   * Download blank tour forms
   */
  async downloadBlankForms() {
    try {
      console.log('📋 Downloading blank tour forms...');
      
      const response = await this.client.api.get('/api/documents/forms/blank', {
        responseType: 'blob'
      });
      
      // In a browser environment, you would create a download link
      // const url = window.URL.createObjectURL(new Blob([response.data]));
      // const link = document.createElement('a');
      // link.href = url;
      // link.download = 'tour-forms.pdf';
      // link.click();
      
      console.log('✅ Blank forms downloaded successfully');
      return response.data;
    } catch (error) {
      console.error('❌ Failed to download forms:', error.message);
      throw error;
    }
  }
  
  /**
   * View my tour registrations
   */
  async viewMyRegistrations() {
    try {
      console.log('📋 Viewing my tour registrations...');
      
      // Get tour events where I'm registered
      const events = await this.client.get('/api/tour-events');
      const myRegistrations = events.filter(event => 
        event.registeredTourists?.includes(this.client.user.userId)
      );
      
      console.log(`🎪 Found ${myRegistrations.length} registrations:`);
      myRegistrations.forEach(event => {
        console.log(`  • ${event.customTourName}`);
        console.log(`    📅 ${event.startDate} to ${event.endDate}`);
        console.log(`    📊 Status: ${event.status}`);
        console.log('');
      });
      
      return myRegistrations;
    } catch (error) {
      console.error('❌ Failed to view registrations:', error.message);
      throw error;
    }
  }
  
  /**
   * Update profile information
   */
  async updateProfile(updates) {
    try {
      console.log('👤 Updating profile...');
      
      const updatedUser = await this.client.put(`/api/users/${this.client.user.userId}`, updates);
      
      console.log('✅ Profile updated successfully');
      console.log(`👤 Name: ${updatedUser.firstName} ${updatedUser.lastName}`);
      console.log(`📧 Email: ${updatedUser.emailAddress}`);
      console.log(`📱 Phone: ${updatedUser.phoneNumber}`);
      
      return updatedUser;
    } catch (error) {
      console.error('❌ Profile update failed:', error.message);
      throw error;
    }
  }
  
  // Helper methods
  getMimeType(fileName) {
    const ext = path.extname(fileName).toLowerCase();
    const mimeTypes = {
      '.pdf': 'application/pdf',
      '.jpg': 'image/jpeg',
      '.jpeg': 'image/jpeg',
      '.png': 'image/png',
      '.doc': 'application/msword',
      '.docx': 'application/vnd.openxmlformats-officedocument.wordprocessingml.document'
    };
    return mimeTypes[ext] || 'application/octet-stream';
  }
  
  formatFileSize(bytes) {
    if (bytes === 0) return '0 Bytes';
    const k = 1024;
    const sizes = ['Bytes', 'KB', 'MB', 'GB'];
    const i = Math.floor(Math.log(bytes) / Math.log(k));
    return parseFloat((bytes / Math.pow(k, i)).toFixed(2)) + ' ' + sizes[i];
  }
}

// Example usage
async function touristExample() {
  const workflow = new TouristWorkflow();
  
  try {
    // Example: Complete tourist workflow
    console.log('=== Tourist Workflow Example ===\n');
    
    // Login (assuming account already exists)
    await workflow.client.login('tourist@example.com', 'password123');
    
    // Browse available tours
    await workflow.browseTourTemplates();
    
    // View available tour events
    const events = await workflow.viewAvailableTourEvents();
    
    if (events.length > 0) {
      // Register for first available tour
      await workflow.registerForTour(events[0].tourEventId);
      
      // View tour schedule
      await workflow.viewTourSchedule(events[0].tourEventId);
    }
    
    // View my documents
    await workflow.viewMyDocuments();
    
    // View my registrations
    await workflow.viewMyRegistrations();
    
    // Logout
    await workflow.client.logout();
    
  } catch (error) {
    console.error('Tourist workflow example failed:', error.message);
  }
}

module.exports = TouristWorkflow;

// Run example if this file is executed directly
if (require.main === module) {
  touristExample();
}

File: src/docs/api-client-examples/python/tourist_workflow.py
"""
Tourist Hub API - Tourist Workflow Examples

This example demonstrates common workflows for Tourist users,
including profile management, tour registration, and document handling.
"""

import os
from typing import List, Dict, Any, Optional
from datetime import datetime
import logging
from basic_client import TouristHubClient

logger = logging.getLogger(__name__)


class TouristWorkflow:
    """Tourist-specific workflow examples."""
    
    def __init__(self, base_url: str = "http://localhost:3000"):
        self.client = TouristHubClient(base_url)
    
    def register_and_setup_profile(self, registration_data: Dict[str, Any]) -> Dict[str, Any]:
        """Complete tourist registration and setup workflow."""
        try:
            logger.info("🚀 Starting tourist registration workflow...")
            
            # 1. Register new tourist account
            logger.info("📝 Registering new account...")
            registration_response = self.client.post('/api/auth/register', {
                **registration_data,
                'userType': 'Tourist'
            })
            logger.info("✅ Registration successful")
            
            # 2. Login with new credentials
            logger.info("🔐 Logging in...")
            self.client.login(registration_data['emailAddress'], registration_data['password'])
            
            # 3. Update profile with additional information
            logger.info("👤 Updating profile...")
            profile_update = {
                'passportNumber': registration_data.get('passportNumber'),
                'dateOfBirth': registration_data.get('dateOfBirth'),
                'gender': registration_data.get('gender')
            }
            
            # Remove None values
            profile_update = {k: v for k, v in profile_update.items() if v is not None}
            
            updated_user = self.client.put(f"/api/users/{self.client.user['userId']}", profile_update)
            logger.info("✅ Profile updated successfully")
            
            return updated_user
        except Exception as e:
            logger.error(f"❌ Registration workflow failed: {e}")
            raise e
    
    def browse_tour_templates(self, filters: Optional[Dict[str, Any]] = None) -> List[Dict[str, Any]]:
        """Browse and search available tour templates."""
        try:
            logger.info("🔍 Browsing tour templates...")
            
            templates = self.client.get('/api/tour-templates', filters or {})
            
            logger.info(f"📋 Found {len(templates)} tour templates:")
            for template in templates:
                logger.info(f"  • {template['templateName']} ({template['type']})")
                logger.info(f"    📅 {template['startDate']} to {template['endDate']}")
                logger.info(f"    📍 {len(template.get('sitesToVisit', []))} sites to visit")
                logger.info("")
            
            return templates
        except Exception as e:
            logger.error(f"❌ Failed to browse templates: {e}")
            raise e
    
    def view_available_tour_events(self, filters: Optional[Dict[str, Any]] = None) -> List[Dict[str, Any]]:
        """View available tour events based on templates."""
        try:
            logger.info("🎯 Viewing available tour events...")
            
            events = self.client.get('/api/tour-events', filters or {})
            
            logger.info(f"🎪 Found {len(events)} available tour events:")
            for event in events:
                logger.info(f"  • {event['customTourName']}")
                logger.info(f"    📅 {event['startDate']} to {event['endDate']}")
                logger.info(f"    👥 {event['remainingTourists']}/{event['numberOfAllowedTourists']} spots available")
                logger.info(f"    🏨 Hotels: {event['place1Hotel']}, {event['place2Hotel']}")
                logger.info(f"    📊 Status: {event['status']}")
                logger.info("")
            
            return events
        except Exception as e:
            logger.error(f"❌ Failed to view tour events: {e}")
            raise e
    
    def register_for_tour(self, tour_event_id: str) -> Dict[str, Any]:
        """Register for a tour event."""
        try:
            logger.info(f"📝 Registering for tour event: {tour_event_id}")
            
            # First, get tour event details
            tour_event = self.client.get(f"/api/tour-events/{tour_event_id}")
            logger.info(f"🎪 Tour: {tour_event['customTourName']}")
            
            # Check availability
            if tour_event['remainingTourists'] <= 0:
                raise ValueError("Tour is fully booked")
            
            if tour_event['status'] != 'Active':
                raise ValueError(f"Tour is not available for registration (Status: {tour_event['status']})")
            
            # Register for the tour
            registration = self.client.post(f"/api/tour-events/{tour_event_id}/register")
            
            logger.info("✅ Registration successful!")
            logger.info(f"📋 Registration Status: {registration['status']}")
            
            if registration['status'] == 'Pending':
                logger.info("⏳ Your registration is pending approval from the provider")
            
            return registration
        except Exception as e:
            logger.error(f"❌ Tour registration failed: {e}")
            raise e
    
    def view_tour_schedule(self, tour_event_id: str) -> List[Dict[str, Any]]:
        """View tour schedule and activities."""
        try:
            logger.info(f"📅 Viewing schedule for tour: {tour_event_id}")
            
            schedule = self.client.get(f"/api/tour-events/{tour_event_id}/schedule")
            
            logger.info("📋 Tour Schedule:")
            for activity in schedule:
                logger.info(f"  📅 {activity['activityDate']} ({activity.get('islamicDate', 'N/A')})")
                logger.info(f"    🕐 {activity['startTime']} - {activity['endTime']}")
                logger.info(f"    🎯 {activity['activityType']}: {activity['description']}")
                logger.info(f"    📍 Location: {activity['location']}")
                if activity.get('webLink'):
                    logger.info(f"    🔗 Link: {activity['webLink']}")
                logger.info("")
            
            return schedule
        except Exception as e:
            logger.error(f"❌ Failed to view schedule: {e}")
            raise e
    
    def upload_document(self, file_path: str, document_type: str, description: str = "") -> Dict[str, Any]:
        """Upload document (passport, ticket, etc.)."""
        try:
            logger.info(f"📄 Uploading {document_type} document...")
            
            # Check if file exists
            if not os.path.exists(file_path):
                raise FileNotFoundError(f"File not found: {file_path}")
            
            upload_data = {
                'type': document_type,
                'description': description
            }
            
            document = self.client.upload_file('/api/documents', file_path, upload_data)
            
            logger.info("✅ Document uploaded successfully")
            logger.info(f"📄 Document ID: {document['documentId']}")
            logger.info(f"📁 File: {document['fileName']}")
            
            return document
        except Exception as e:
            logger.error(f"❌ Document upload failed: {e}")
            raise e
    
    def view_my_documents(self) -> List[Dict[str, Any]]:
        """View uploaded documents."""
        try:
            logger.info("📂 Viewing my documents...")
            
            documents = self.client.get('/api/documents')
            
            logger.info(f"📋 Found {len(documents)} documents:")
            for doc in documents:
                logger.info(f"  📄 {doc['fileName']} ({doc['type']})")
                logger.info(f"    📝 {doc.get('description', 'No description')}")
                logger.info(f"    📅 Uploaded: {datetime.fromisoformat(doc['uploadDate'].replace('Z', '+00:00')).strftime('%Y-%m-%d')}")
                logger.info(f"    💾 Size: {self._format_file_size(doc['fileSize'])}")
                logger.info("")
            
            return documents
        except Exception as e:
            logger.error(f"❌ Failed to view documents: {e}")
            raise e
    
    def download_blank_forms(self, save_path: str = "tour-forms.pdf") -> str:
        """Download blank tour forms."""
        try:
            logger.info("📋 Downloading blank tour forms...")
            
            # Make request with stream=True for file download
            import requests
            headers = self.client._get_headers()
            url = f"{self.client.base_url}/api/documents/forms/blank"
            
            response = requests.get(url, headers=headers, stream=True)
            response.raise_for_status()
            
            with open(save_path, 'wb') as f:
                for chunk in response.iter_content(chunk_size=8192):
                    f.write(chunk)
            
            logger.info(f"✅ Blank forms downloaded to: {save_path}")
            return save_path
        except Exception as e:
            logger.error(f"❌ Failed to download forms: {e}")
            raise e
    
    def view_my_registrations(self) -> List[Dict[str, Any]]:
        """View my tour registrations."""
        try:
            logger.info("📋 Viewing my tour registrations...")
            
            # Get tour events where I'm registered
            events = self.client.get('/api/tour-events')
            my_registrations = [
                event for event in events 
                if self.client.user['userId'] in event.get('registeredTourists', [])
            ]
            
            logger.info(f"🎪 Found {len(my_registrations)} registrations:")
            for event in my_registrations:
                logger.info(f"  • {event['customTourName']}")
                logger.info(f"    📅 {event['startDate']} to {event['endDate']}")
                logger.info(f"    📊 Status: {event['status']}")
                logger.info("")
            
            return my_registrations
        except Exception as e:
            logger.error(f"❌ Failed to view registrations: {e}")
            raise e
    
    def update_profile(self, updates: Dict[str, Any]) -> Dict[str, Any]:
        """Update profile information."""
        try:
            logger.info("👤 Updating profile...")
            
            updated_user = self.client.put(f"/api/users/{self.client.user['userId']}", updates)
            
            logger.info("✅ Profile updated successfully")
            logger.info(f"👤 Name: {updated_user['firstName']} {updated_user['lastName']}")
            logger.info(f"📧 Email: {updated_user['emailAddress']}")
            logger.info(f"📱 Phone: {updated_user['phoneNumber']}")
            
            return updated_user
        except Exception as e:
            logger.error(f"❌ Profile update failed: {e}")
            raise e
    
    def _format_file_size(self, bytes_size: int) -> str:
        """Format file size in human readable format."""
        if bytes_size == 0:
            return "0 Bytes"
        
        k = 1024
        sizes = ["Bytes", "KB", "MB", "GB"]
        i = 0
        
        while bytes_size >= k and i < len(sizes) - 1:
            bytes_size /= k
            i += 1
        
        return f"{bytes_size:.2f} {sizes[i]}"


def tourist_example():
    """Example usage of tourist workflow."""
    workflow = TouristWorkflow()
    
    try:
        logger.info("=== Tourist Workflow Example ===\n")
        
        # Login (assuming account already exists)
        workflow.client.login('tourist@example.com', 'password123')
        
        # Browse available tours
        workflow.browse_tour_templates()
        
        # View available tour events
        events = workflow.view_available_tour_events()
        
        if events:
            # Register for first available tour
            workflow.register_for_tour(events[0]['tourEventId'])
            
            # View tour schedule
            workflow.view_tour_schedule(events[0]['tourEventId'])
        
        # View my documents
        workflow.view_my_documents()
        
        # View my registrations
        workflow.view_my_registrations()
        
        # Logout
        workflow.client.logout()
        
    except Exception as e:
        logger.error(f"Tourist workflow example failed: {e}")


if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)
    tourist_example()